postgis/lwgeom_geos.c

```diff
@@ -3256,6 +3263,142 @@ POSTGIS2GEOS(GSERIALIZED *pglwgeom)
 	return ret;
 }
 
+uint32_t array_nelems_not_null(ArrayType* array) {
+    ArrayIterator iterator;
+    Datum value;
+    bool isnull;
+    uint32_t nelems_not_null = 0;
+
+#if POSTGIS_PGSQL_VERSION >= 95	
+	iterator = array_create_iterator(array, 0, NULL);
+#else
+	iterator = array_create_iterator(array, 0);
+#endif
+	while(array_iterate(iterator, &value, &isnull) )
+	{
+        if (!isnull)
+        {
+            nelems_not_null++;
+        }
+    }
+    array_free_iterator(iterator);
+
+    return nelems_not_null;
+}
+
+/* ARRAY2GEOS: Converts the non-null elements of a Postgres array into a LWGEOM* array */
+LWGEOM** ARRAY2LWGEOM(ArrayType* array, uint32_t nelems,  int* is3d, int* srid)
+{
+    ArrayIterator iterator;
+    Datum value;
+    bool isnull;
+    bool gotsrid = false;
+    uint32_t i = 0;
+
+	LWGEOM** lw_geoms = palloc(nelems * sizeof(LWGEOM*));
+
+#if POSTGIS_PGSQL_VERSION >= 95
+    iterator = array_create_iterator(array, 0, NULL);
+#else
+    iterator = array_create_iterator(array, 0);
+#endif
+
+    while(array_iterate(iterator, &value, &isnull))
+	{
+		GSERIALIZED *geom = (GSERIALIZED*) DatumGetPointer(value);
+
+        if (isnull)
+        {
+            continue;
+        }
+
+		*is3d = *is3d || gserialized_has_z(geom);
+
+		lw_geoms[i] = lwgeom_from_gserialized(geom);
+		if (!lw_geoms[i]) /* error in creation */
+		{
+			lwpgerror("Geometry deserializing geometry");
+			return NULL;
+		}
+		if (!gotsrid)
+		{
+            gotsrid = true;
+			*srid = gserialized_get_srid(geom);
+		}
+		else if (*srid != gserialized_get_srid(geom))
+		{
+            error_if_srid_mismatch(*srid, gserialized_get_srid(geom));
+			return NULL;
+		}
+
+        i++;
+	}
+
+	return lw_geoms;
+}
+
+/* ARRAY2GEOS: Converts the non-null elements of a Postgres array into a GEOSGeometry* array */
+GEOSGeometry** ARRAY2GEOS(ArrayType* array, uint32_t nelems, int* is3d, int* srid)
+{
+    ArrayIterator iterator;
+    Datum value;
+    bool isnull;
+    bool gotsrid = false;
+    uint32_t i = 0;
+
+	GEOSGeometry** geos_geoms = palloc(nelems * sizeof(GEOSGeometry*));
+
+#if POSTGIS_PGSQL_VERSION >= 95
+    iterator = array_create_iterator(array, 0, NULL);
+#else
+    iterator = array_create_iterator(array, 0);
+#endif
+
+    while(array_iterate(iterator, &value, &isnull))
+	{
+        GSERIALIZED *geom = (GSERIALIZED*) DatumGetPointer(value);
+
+        if (isnull)
+        {
+            continue;
+        }
+
+		*is3d = *is3d || gserialized_has_z(geom);
+
+		geos_geoms[i] = (GEOSGeometry*) POSTGIS2GEOS(geom);
+		if (!geos_geoms[i])   /* exception thrown at construction */
+		{
+            lwpgerror("Geometry could not be converted to GEOS");
+
+            while (i >= 0)
```

Comment:
This looks like an infinite loop to me... 'i' is declared unsigned (also, Don't see it initialized to anything other than 0)